### DISCUSSION TITLE
Are there tornado listeners?

### DISCUSSION THREAD

**Discussion by zealbell** [beehive-lab/TornadoVM#224] [Q&A]
I want some code triggered immediately my `TaskGraph` finishes execution i.e. at the point `transferToHost(…)` returns, So I am just wondering if Tornado has some sorta listener interface because what I am doing currently for the piece of code I want triggered is: blocking my main thread until the parameter I passed into `transferToHost(…)` is non null before running it.

**Comment by jjfumero** [comment]
`transferToHost` and `transferToDevice` just annotate the Java objects that need to be transferred. There is no data transfer at that point. Only when the `execute` method is invoked, the whole task-graph is executed on the accelerator.

**Reply by zealbell** [reply]
Yes I am aware. I am actually referring to the point say at runtime where all accelerated computations are done, I’d like some code to be executed 
immediately or right away (Asper at the point actual data transfer has taken place after execution I’d like to trigger some events immediately).

**Reply by jjfumero** [reply]
What do you mean by right-away? Can you write an example? 

To answer your first question. Currently, the TornadoVM API does not have any aync call (or listeners)

**Reply by zealbell** [reply]
```java
TaskGraph taskGraph = new TaskGraph("s0") 
                .lockObjectsInMemory(matrixA, matrixB, matrixC) 
                .transferToDevice(DataTransferMode.FIRST_EXECUTION, matrixA, matrixB) 
                .task("t0", MatrixMultiplication2D::matrixMultiplication, matrixA, matrixB, matrixC, size) 
                .transferToHost(matrixC);//I believe at some point after execution this call enables tornado transfer the result of the multiplication from the acceleration device to the jvm host device object by storing within matrixC, is this correct?

  
//something along the lines of
taskGraph.setListener(new TransferToHostListener(){

        @override
        public void beforeTransfer(){
        
        }
        
        @override
        public void onTransfer(){
        
        }

        @override
        public void afterTransfer(){
               doSomeStuff(); //trigger something
        }

  };


taskGraph.execute();//a listener previously since execute() doesn’t block this thread or does it?
doSomeOtherStuff(); //does the execute() call block this thread until the acceleration device is done (then there would be no point of having a listener  since doSomeStuff() can come just before doSomeOtherStuff() i.e. immediately after execute())?


```

**Reply by zealbell** [reply]
Hi, this question is a bit mistaken. I went over [this ](https://www.infoq.com/articles/tornadovm-java-gpu-fpga/) again and I realized `execute()` is a blocking call. Apologies

**Reply by jjfumero** [reply]
Yes, that is what I meant. In the future, we want also to include a non-blocking `execute` call. But for now, it is blocking.
